linearfold.go

```diff
@@ -0,0 +1,1361 @@
+		for l2 := SINGLE_MIN_LEN; l2 <= SINGLE_MAX_LEN; l2++ {
+			if l1 == 0 && l2 == 0 {
+				continue
+			} else if l1 == 0 {
+				cache_single[l1][l2] += bulge_length[l2]
+			} else if l2 == 0 {
+				cache_single[l1][l2] += bulge_length[l1]
+			} else {
+				// internal
+				cache_single[l1][l2] += internal_length[Min(l1+l2, INTERNAL_MAX_LEN)]
+
+				// internal explicit
+				if l1 <= EXPLICIT_MAX_LEN && l2 <= EXPLICIT_MAX_LEN {
+					if l1 <= l2 {
+						cache_single[l1][l2] += internal_explicit[l1*EXPLICIT_MAX_LEN+l2]
+					} else {
+						cache_single[l1][l2] += internal_explicit[l2*EXPLICIT_MAX_LEN+l1]
+					}
+				}
+
+				if l1 == l2 {
+					// internal symmetry
+					cache_single[l1][l2] += internal_symmetric_length[Min(l1, SYMMETRIC_MAX_LEN)]
+				} else {
+					// internal asymmetry
+					var diff int = l1 - l2
+					if diff < 0 {
+						diff = -diff
+					}
+					cache_single[l1][l2] += internal_asymmetry[Min(diff, ASYMMETRY_MAX_LEN)]
+				}
+			}
+
+		}
+	}
+}
+
+// What is manner?
+const (
+	MANNER_NONE                  = iota // 0: empty
+	MANNER_H                            // 1: hairpin candidate
+	MANNER_HAIRPIN                      // 2: hairpin
+	MANNER_SINGLE                       // 3: single
+	MANNER_HELIX                        // 4: helix
+	MANNER_MULTI                        // 5: multi = ..M2. [30 restriction on the left and jump on the right]
+	MANNER_MULTI_eq_MULTI_plus_U        // 6: multi = multi + U
+	MANNER_P_eq_MULTI                   // 7: P = (multi)
+	MANNER_M2_eq_M_plus_P               // 8: M2 = M + P
+	MANNER_M_eq_M2                      // 9: M = M2
+	MANNER_M_eq_M_plus_U                // 10: M = M + U
+	MANNER_M_eq_P                       // 11: M = P
+	/* MANNER_C_eq_U, */
+	/* MANNER_C_eq_P, */
+	MANNER_C_eq_C_plus_U // 12: C = C + U
+	MANNER_C_eq_C_plus_P // 13: C = C + P
+)
+
+//
+type State struct {
+	score float64
+	// age    int
+	manner int
+	// trace is a union in the original code. Need to be aware of this!
+	trace struct {
+		split    int
+		paddings struct {
+			l1 rune
+			l2 int
+		}
+	}
+}
+
+func NewState() *State {
+	return &State{score: VALUE_MIN, manner: MANNER_NONE}
+}
+
+func update_if_better(s *State, newscore float64, manner int) {
+	if s.score < newscore {
+		s.Set(newscore, manner)
+	}
+}
+func update_if_better2(s *State, newscore float64, manner int, l1 rune, l2 int) {
+	if s.score < newscore || s.manner == MANNER_NONE {
+		s.Set2(newscore, manner, l1, l2)
+	}
+}
+
+func update_if_better3(s *State, newscore float64, manner int, split int) {
+	if s.score < newscore || s.manner == MANNER_NONE {
+		s.Set3(newscore, manner, split)
+	}
+}
+
+func (s *State) Set(score float64, manner int) {
+	s.score = score
+	s.manner = manner
+}
+
+func (s *State) Set2(score float64, manner int, l1 rune, l2 int) {
+	s.score = score
+	s.manner = manner
+	s.trace.paddings.l1 = l1
+	s.trace.paddings.l2 = l2
+}
+
+func (s *State) Set3(score float64, manner int, split int) {
```

Comment:
Why are there 3 set state methods and why are they are setting slightly different things that are unnamed?

Reply:
Yes.

Reply:
![Spanish laughing guy collage](https://user-images.githubusercontent.com/7042260/118372857-74754200-b568-11eb-92d9-e19be8ce066a.jpg)